processor/src/main/java/org/mapstruct/ap/internal/model/source/selector/LifecycleOverloadDeduplicateSelector.java

```diff
@@ -0,0 +1,131 @@
+ * except for the target type's inheritance relationship.
+ */
+public class LifecycleOverloadDeduplicateSelector implements MethodSelector {
+    @Override
+    public <T extends Method> List<SelectedMethod<T>> getMatchingMethods(List<SelectedMethod<T>> methods,
+                                                                         SelectionContext context) {
+        if ( !context.getSelectionCriteria().isLifecycleCallbackRequired() ) {
+            return methods;
+        }
+        List<List<SelectedMethod<T>>> methodSignatureGroups = new ArrayList<>(
+            methods.stream()
+                .collect( Collectors.groupingBy(
+                    LifecycleOverloadDeduplicateSelector::buildSignatureKey,
+                    LinkedHashMap::new,
+                    Collectors.toList()
+                ) )
+                .values()
+        );
+        List<SelectedMethod<T>> deduplicatedMethods = new ArrayList<>();
+        for ( List<SelectedMethod<T>> signatureGroup : methodSignatureGroups ) {
+            if ( signatureGroup.size() == 1 ) {
+                deduplicatedMethods.add( signatureGroup.get( 0 ) );
+                continue;
+            }
+            SelectedMethod<T> bestInheritanceMethod = signatureGroup.get( 0 );
+            for ( int i = 1; i < signatureGroup.size(); i++ ) {
+                SelectedMethod<T> candidateMethod = signatureGroup.get( i );
+                if ( isInheritanceBetter( candidateMethod, bestInheritanceMethod ) ) {
+                    bestInheritanceMethod = candidateMethod;
+                }
+            }
+            deduplicatedMethods.add( bestInheritanceMethod );
+        }
+        return deduplicatedMethods;
+    }
+
+    /**
+     * Builds a grouping key for a method based on its defining type,
+     * method name, and a detailed breakdown of each parameter binding.
+     * <p>
+     * The key consists of:
+     * <ul>
+     *   <li>The type that defines the method</li>
+     *   <li>The method name</li>
+     *   <li>parameter bindings</li>
+     * </ul>
+     * This ensures that methods are grouped together only if all these aspects match,
+     * except for differences in type hierarchy, which are handled separately.
+     */
+    private static <T extends Method> List<Object> buildSignatureKey(SelectedMethod<T> method) {
+        List<Object> key = new ArrayList<>();
+        key.add( method.getMethod().getDefiningType() );
+        key.add( method.getMethod().getName() );
+        for ( ParameterBinding binding : method.getParameterBindings() ) {
+            key.add( binding.getType() );
+            key.add( binding.getVariableName() );
+            key.add( binding.isMappingTarget() );
+            key.add( binding.isMappingContext() );
+            key.add( binding.isTargetType() );
+            key.add( binding.isSourcePropertyName() );
+            key.add( binding.isTargetPropertyName() );
```

Comment:
Why should these elements be part of the signature?

Reply:
This is not necessary.  The type and variable name are sufficient to determine the parameter signature.

Reply:
Cool thanks, Shall we remove those then? And just so that I understand properly, when you say variable name, it is the name of something in the mapping method, not the name of a parameter in the lifecycle method, right?

Reply:
> Cool thanks, Shall we remove those then? And just so that I understand properly, when you say variable name, it is the name of something in the mapping method, not the name of a parameter in the lifecycle method, right?
Of course.

> when you say variable name, it is the name of something in the mapping method, not the name of a parameter in the lifecycle method, right?
Yes.